core/matrix/fbcsr.cpp

```diff
@@ -0,0 +1,513 @@
+
+
+namespace gko {
+namespace matrix {
+namespace fbcsr {
+
+
+GKO_REGISTER_OPERATION(spmv, fbcsr::spmv);
+GKO_REGISTER_OPERATION(advanced_spmv, fbcsr::advanced_spmv);
+GKO_REGISTER_OPERATION(convert_to_csr, fbcsr::convert_to_csr);
+GKO_REGISTER_OPERATION(convert_to_dense, fbcsr::convert_to_dense);
+GKO_REGISTER_OPERATION(transpose, fbcsr::transpose);
+GKO_REGISTER_OPERATION(conj_transpose, fbcsr::conj_transpose);
+GKO_REGISTER_OPERATION(calculate_max_nnz_per_row,
+                       fbcsr::calculate_max_nnz_per_row);
+GKO_REGISTER_OPERATION(calculate_nonzeros_per_row,
+                       fbcsr::calculate_nonzeros_per_row);
+GKO_REGISTER_OPERATION(is_sorted_by_column_index,
+                       fbcsr::is_sorted_by_column_index);
+GKO_REGISTER_OPERATION(extract_diagonal, fbcsr::extract_diagonal);
+GKO_REGISTER_OPERATION(fill_array, components::fill_array);
+GKO_REGISTER_OPERATION(inplace_absolute_array,
+                       components::inplace_absolute_array);
+GKO_REGISTER_OPERATION(outplace_absolute_array,
+                       components::outplace_absolute_array);
+
+
+}  // namespace fbcsr
+
+
+namespace detail {
+
+
+/**
+ * A lightweight dynamic block type on the host
+ *
+ * @internal Currently used only while reading a FBCSR matrix from matrix_data.
+ *
+ * @tparam ValueType  The numeric type of entries of the block
+ */
+template <typename ValueType>
+class DenseBlock final {
+public:
+    using value_type = ValueType;
+
+    DenseBlock() = default;
+
+    DenseBlock(const int num_rows, const int num_cols)
+        : nrows_{num_rows}, ncols_{num_cols}, vals_(num_rows * num_cols)
+    {}
+
+    value_type &at(const int row, const int col)
+    {
+        return vals_[row * ncols_ + col];
+    }
+
+    const value_type &at(const int row, const int col) const
+    {
+        return vals_[row * ncols_ + col];
+    }
+
+    value_type &operator()(const int row, const int col)
+    {
+        return at(row, col);
+    }
+
+    const value_type &operator()(const int row, const int col) const
+    {
+        return at(row, col);
+    }
+
+    int size() const { return nrows_ * ncols_; }
+
+    void resize(const int nrows, const int ncols)
+    {
+        vals_.resize(nrows * ncols);
+        nrows_ = nrows;
+        ncols_ = ncols;
+    }
+
+    void zero()
+    {
+        std::fill(vals_.begin(), vals_.end(), static_cast<value_type>(0));
+    }
+
+private:
+    int nrows_ = 0;
+    int ncols_ = 0;
+    std::vector<value_type> vals_;
+};
+
+
+}  // namespace detail
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::apply_impl(const LinOp *const b,
+                                             LinOp *const x) const
+{
+    using Dense = Dense<ValueType>;
+    if (auto b_fbcsr = dynamic_cast<const Fbcsr<ValueType, IndexType> *>(b)) {
+        // if b is a FBCSR matrix, we compute a SpGeMM
+        throw NotImplemented(__FILE__, __LINE__, "SpGeMM for Fbcsr");
```

Comment:
Can you maybe just change this to a `GKO_NOT_IMPLEMENTED` and for the functions below ? I understand that you want to add a comment on where this was called, but I think it would be better for uniformity we could pass a message to `GKO_NOT_IMPLEMENTED` and call `GKO_NOT_IMPLEMENTED(message)` instead.

Reply:
In this case, I'm not satisfied with just getting the line number and the function name as `apply_impl`, so I want a more specific message, yes. You want me to add an optional message argument to the `GKO_NOT_IMPLEMENTED` macro?

Reply:
Yes, if you can make it an optional message argument to `GKO_NOT_IMPLEMENTED`, that would be good.